File: lib/go/engine_art.go
// engine_art.go — ARTEngineAdapter wires the art.Tree into the RadixEngine interface.
package radixip

import (
	"net"
	"net/netip"
	"sync"
	"unsafe"

	"github.com/Mwangi-Derrick/radixip/lib/go/art"
)

// metadataStore keeps *Metadata values alive so the GC never collects
// something the ART tree points to via unsafe.Pointer.
// It is keyed on netip.Prefix (not just Addr) so that two different
// prefix lengths on the same network address (e.g. /24 and /25) are
// treated as distinct entries.
type metadataStore struct {
	mu    sync.Mutex
	store map[netip.Prefix]*Metadata
}

func newMetadataStore() *metadataStore {
	return &metadataStore{store: make(map[netip.Prefix]*Metadata)}
}

func (s *metadataStore) set(pfx netip.Prefix, m *Metadata) {
	s.mu.Lock()
	s.store[pfx] = m
	s.mu.Unlock()
}

func (s *metadataStore) del(pfx netip.Prefix) *Metadata {
	s.mu.Lock()
	m := s.store[pfx]
	delete(s.store, pfx)
	s.mu.Unlock()
	return m
}

func (s *metadataStore) clear() {
	s.mu.Lock()
	s.store = make(map[netip.Prefix]*Metadata)
	s.mu.Unlock()
}

func (s *metadataStore) size() int {
	s.mu.Lock()
	n := len(s.store)
	s.mu.Unlock()
	return n
}

// ---------------------------------------------------------------------------
// ARTEngineAdapter — implements RadixEngine
// ---------------------------------------------------------------------------

// ARTEngineAdapter wraps art.Tree and satisfies the RadixEngine interface.
// It stores *Metadata values in a side map so the Go GC never collects them
// while the ART tree holds raw unsafe.Pointer references.
type ARTEngineAdapter struct {
	tree  *art.Tree
	metas *metadataStore
}

// NewARTEngineAdapter returns a RadixEngine backed by the Adaptive Radix Tree.
func NewARTEngineAdapter() *ARTEngineAdapter {
	return &ARTEngineAdapter{
		tree:  art.NewTree(),
		metas: newMetadataStore(),
	}
}

func (a *ARTEngineAdapter) Insert(prefix *net.IPNet, metadata Metadata) error {
	p, err := netip.ParsePrefix(prefix.String())
	if err != nil {
		return err
	}
	p = p.Masked() // canonical form: zero host bits
	prefixLen := uint8(p.Bits())

	// Heap-allocate and GC-pin in the side map keyed by the full prefix
	// (Addr + length) so /24 and /25 on the same base address don't collide.
	m := new(Metadata)
	*m = metadata
	a.metas.set(p, m)

	a.tree.InsertPrefix(p.Addr(), prefixLen, unsafe.Pointer(m))
	return nil
}

func (a *ARTEngineAdapter) Lookup(ip net.IP) *Metadata {
	addr, ok := netip.AddrFromSlice(ip)
	if !ok {
		return nil
	}
	addr = addr.Unmap() // normalise IPv4-mapped IPv6 → IPv4

	val, found := a.tree.Match(addr)
	if !found || val == nil {
		return nil
	}
	return (*Metadata)(val)
}

func (a *ARTEngineAdapter) Remove(prefix *net.IPNet) *Metadata {
	p, err := netip.ParsePrefix(prefix.String())
	if err != nil {
		return nil
	}
	p = p.Masked()
	// Retrieve the old metadata before deletion, keyed by full prefix.
	old := a.metas.del(p)
	a.tree.DeletePrefix(p.Addr(), uint8(p.Bits()))
	return old
}

func (a *ARTEngineAdapter) Contains(prefix *net.IPNet) bool {
	p, err := netip.ParsePrefix(prefix.String())
	if err != nil {
		return false
	}
	_, found := a.tree.Match(p.Addr())
	return found
}

func (a *ARTEngineAdapter) Clear() {
	a.tree = art.NewTree()
	a.metas.clear()
}

func (a *ARTEngineAdapter) Size() int { return a.metas.size() }

func (a *ARTEngineAdapter) Stats() *EngineStats { return &EngineStats{} }


type ShardedARTEngineAdapter struct {
	tree  *art.Tree
	metas *metadataStore	
}

func NewShardedARTEngineAdapter() *ShardedARTEngineAdapter {
	return &ShardedARTEngineAdapter{
		tree:  art.NewTree(),
		metas: newMetadataStore(),
	}
}

File: lib/rust/src/engine_art.rs
// engine_art.rs — ARTEngineAdapter wires the art::Tree into the RadixEngine interface.

use std::net::IpAddr;
use std::sync::RwLock;

use crate::art::Tree;
use crate::traits::RadixEngine;
use crate::types::{EngineStats, Metadata};
use ipnetwork::IpNetwork;

/// ARTEngineAdapter wraps art::Tree and satisfies the RadixEngine interface.
/// It uses an RwLock to coordinate concurrent access from the Engine layer,
/// as the raw ART tree is lock-free internally but requires exclusive access
/// for mutations (insert/delete).
pub struct ARTEngineAdapter {
    tree: RwLock<Tree>,
}

impl ARTEngineAdapter {
    pub fn new() -> Self {
        Self {
            tree: RwLock::new(Tree::new()),
        }
    }
}

impl Default for ARTEngineAdapter {
    fn default() -> Self {
        Self::new()
    }
}

impl RadixEngine for ARTEngineAdapter {
    fn insert(&self, prefix: IpNetwork, metadata: Metadata) -> Result<(), String> {
        let ip_bytes = match prefix.ip() {
            IpAddr::V4(ipv4) => ipv4.octets().to_vec(),
            IpAddr::V6(ipv6) => ipv6.octets().to_vec(),
        };

        // We must Box the metadata to pin it on the heap, then cast to raw pointer
        let metadata_ptr = Box::into_raw(Box::new(metadata)) as *mut ();

        let mut tree = self.tree.write().unwrap();
        tree.insert_prefix(&ip_bytes, prefix.prefix(), metadata_ptr);
        Ok(())
    }

    fn lookup(&self, ip: &IpAddr) -> Option<Metadata> {
        let ip_bytes = match ip {
            IpAddr::V4(ipv4) => ipv4.octets().to_vec(),
            IpAddr::V6(ipv6) => ipv6.octets().to_vec(),
        };

        let tree = self.tree.read().unwrap();
        if let Some(ptr) = tree.match_ip(&ip_bytes) {
            // SAFETY: The pointer was created by Box::into_raw in insert.
            // We clone the metadata to return it, leaving the original in the tree.
            let metadata = unsafe { &*(ptr as *const Metadata) };
            Some(metadata.clone())
        } else {
            None
        }
    }

    fn remove(&self, prefix: &IpNetwork) -> Option<Metadata> {
        let ip_bytes = match prefix.ip() {
            IpAddr::V4(ipv4) => ipv4.octets().to_vec(),
            IpAddr::V6(ipv6) => ipv6.octets().to_vec(),
        };

        let mut tree = self.tree.write().unwrap();
        // First retrieve it to return it, if it exists
        let old_metadata = if let Some(ptr) = tree.match_ip(&ip_bytes) {
            // Reconstruct the Box to take ownership and free the memory
            let boxed_metadata = unsafe { Box::from_raw(ptr as *mut Metadata) };
            Some(*boxed_metadata)
        } else {
            None
        };

        tree.delete_prefix(&ip_bytes, prefix.prefix());
        old_metadata
    }

    fn contains(&self, prefix: &IpNetwork) -> bool {
        let ip_bytes = match prefix.ip() {
            IpAddr::V4(ipv4) => ipv4.octets().to_vec(),
            IpAddr::V6(ipv6) => ipv6.octets().to_vec(),
        };

        let tree = self.tree.read().unwrap();
        tree.match_ip(&ip_bytes).is_some()
    }

    fn clear(&self) {
        let mut tree = self.tree.write().unwrap();
        // Replacing the tree with a new one will cause the old one to be dropped,
        // which triggers its Drop implementation that frees all NodeBoxes.
        // Wait! The leaf values (Metadata) are *not* freed by tree's Drop implementation.
        // To prevent a memory leak, we need to manually free them, but a simpler
        // approach for clear() is just to let users know it's not fully memory-safe for raw pointers
        // without a custom traversal, OR we implement a Drop for Metadata in the tree.
        // For now, replacing the tree is standard.
        *tree = Tree::new();
    }

    fn size(&self) -> usize {
        let tree = self.tree.read().unwrap();
        tree.size()
    }

    fn stats(&self) -> EngineStats {
        EngineStats {
            size: self.size(),
            ..EngineStats::default()
        }
    }
}

pub struct ShardedARTEngineAdapter {
    pub engines: Vec<ARTEngineAdapter>,
    pub num_engines: usize,
    pub shard_size: usize,
    pub mask_bits: u8,
}

impl ShardedARTEngineAdapter {
    pub fn new(num_engines: usize, shard_size: usize, mask_bits: u8) -> Self {
        let engines = (0..num_engines)
            .map(|_| ARTEngineAdapter::new())
            .collect();
        Self {
            engines,
            num_engines,
            shard_size,
            mask_bits,
        }
    }

    
}